drivers/wasapi/audio_driver_wasapi.cpp

```diff
@@ -732,241 +814,304 @@ void AudioDriverWASAPI::write_sample(WORD format_tag, int bits_per_sample, BYTE
+			Error err = finish_input_device();
 			if (err != OK) {
-				ERR_PRINT("WASAPI: finish_output_device error");
+				ERR_PRINT("WASAPI: finish_input_device error");
 			}
+		} else if (hr != S_OK) {
+			ERR_PRINT("WASAPI: GetNextPacketSize error");
+		}
+	}
+}
 
-			default_output_device_changed = false;
+// Helper function to handle output device changes
+void AudioDriverWASAPI::handle_output_device_changes() {
+	// If we're using the Default output device and it changed finish it so we'll re-init the output device
+	if (audio_output.device_name == "Default" && default_output_device_changed) {
+		Error err = finish_output_device();
+		if (err != OK) {
+			ERR_PRINT("WASAPI: finish_output_device error");
 		}
 
-		// User selected a new output device, finish the current one so we'll init the new output device
-		if (ad->audio_output.device_name != ad->audio_output.new_device) {
-			ad->audio_output.device_name = ad->audio_output.new_device;
-			Error err = ad->finish_output_device();
-			if (err != OK) {
-				ERR_PRINT("WASAPI: finish_output_device error");
+		default_output_device_changed = false;
+	}
+
+	// User selected a new output device, finish the current one so we'll init the new output device
+	if (audio_output.device_name != audio_output.new_device) {
+		audio_output.device_name = audio_output.new_device;
+		Error err = finish_output_device();
+		if (err != OK) {
+			ERR_PRINT("WASAPI: finish_output_device error");
+		}
+	}
+
+	// Reinitialize device if it's not available (either due to invalidation or user change)
+	if (!audio_output.audio_client) {
+		if (output_reinit_countdown < 1) {
+			Error err = init_output_device(true);
+			if (err == OK) {
+				start();
+				// Reset countdown on successful initialization
+				output_reinit_countdown = 0;
+			} else {
+				// Only increment countdown on failure to allow immediate retry on success
+				output_reinit_countdown = 1000;
 			}
+		} else {
+			output_reinit_countdown--;
+		}
+	}
+}
+
+// Helper function to handle input device changes
+void AudioDriverWASAPI::handle_input_device_changes() {
+	if (audio_input.active.is_set()) {
+		bool reinitDevice = false;
+
+		// If we're using the Default input device and it changed finish it so we'll re-init the input device
+		if (audio_input.device_name == "Default" && default_input_device_changed) {
+			reinitDevice = audio_input.active.is_set();
+			default_input_device_changed = false;
+		}
+		// User selected a new input device, finish the current one so we'll init the new input device
+		if (audio_input.device_name != audio_input.new_device) {
+			audio_input.device_name = audio_input.new_device;
+			reinitDevice = audio_input.active.is_set();
 		}
 
-		if (!ad->audio_output.audio_client) {
-			if (output_reinit_countdown < 1) {
-				Error err = ad->init_output_device(true);
+		// Reinitialize device if it's not available (either due to invalidation or user change)
+		if (reinitDevice) {
+			if (input_reinit_countdown < 1) {
```

Comment:
this code no longer calls finish_input_device. Why did this move from here to the middle of the read loop: there might be cases other than `AUDCLNT_E_DEVICE_INVALIDATED` in which the user wishes to finish the current input device and switch.